README.md
# thaidate

## NAME
**thaidate** - accepts 3 numbers, year, month and day, and outputs
a date string in Thai language with the month using the Thai name.


## SYNOPSIS

**thaidate** [option] **y m d**.

## DESCRIPTION

**thaidate** accepts as input, 3 numbers which by default must be in
order of 4 digit year, month number [1..12], and day [1..*maxday*],
where *maxday* is calculated by month and year. You may alter the
default order of input fields using options.

By default the **output** order is **d m y** and also by default Latin
digits are used for the numbers, **d** and **y**.

The Thai calendar year is 543 years ahead of the normally accepted
Western calendar. This increment is added to the output year.

## OPTIONS

+ **-h**, **--help**  
print help information and exit.

+ **-m**, **--mdy**  
order the **input** date fields as month, day and year.

+ **-d**, **--dmy**  
order the **input** date fields as day, month and year.

+ **-y**, **--ymd**  
order the **input** date fields as year, month and day. This is
actually the default.

+ **-Y**, **--YMD**  
orders the **output** fields as year, month, day. This the officially
recognised Thai date format. But in spite of this, the _default_
output ordering is more usual.

+ **-D**, **--DMY**  
orders the **output** fields as day, month, year. This the most common
Thai date format, and it is the default.

+ **-r**, **--real-thai**  
Latin digits are used very commonly in Thailand in many situations,
no doubt thanks to the use the ubiquitous electronic calculator. So,
by default the output of this program uses Latin digits for the day and
year fields. Choosing this option will output Thai digits for numbers
instead.

Example: **thaidate** --real-thai 2018 1 31  
**๓๑ มกราคม ๒๕๖๑**


## BUGS/FEATURES  
Though within Thai writing spaces are not as commonly used as in
English, the writing is by no means space free. Because I *seem* to
recall having seen spaces in Thai dates, I have chosen to follow my
belief. I shall certainly change this on viewing evidence to the
contrary.

## AUTHOR  
Robert L Parker Bob.Parker@programmer.net

## SEE ALSO  
**date** (1)



NEWS
NEWS for thaidate


gopt.c
/*      gopt.c
 *
 *  Copyright 2017 Robert L (Bob) Parker rlp1938@gmail.com
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *  MA 02110-1301, USA.
*/

#include "str.h"
#include "files.h"
#include "gopt.h"


options_t process_options(int argc, char **argv)
{
	optstring = ":hmdyYDr";

	/* declare and set defaults for local variables. */

	/* set up defaults for opt vars. */
	options_t opts = {0};	// assumes defaults all 0/NULL
	opts.o_y = 1;	// ymd is default input format.
	opts.o_D = 1;	// DMY is default output format.

	int c;

	while(1) {
		int this_option_optind = optind ? optind : 1;
		int option_index = 0;
		static struct option long_options[] = {
		{"help",		0,	0,	'h'},
		{"mdy",			0,	0,	'm'},
		{"dmy",			0,	0,	'd'},
		{"ymd",			0,	0,	'y'},
		{"YMD",			0,	0,	'Y'},
		{"DMY",			0,	0,	'D'},
		{"real-thai",	0,	0,	'r'},
		{0,	0,	0,	0 }
		};

		c = getopt_long(argc, argv, optstring,
                        long_options, &option_index);

		if (c == -1)
			break;

		switch (c) {
		case 0:
			switch (option_index) {
			} // switch()
		break;
		case 'h':
		dohelp(0);
		break;
		/* input options */
		case 'm':	// stupid USA format input.
			opts.o_m = 1;
			opts.o_y = 0;
			opts.o_d = 0;
			break;
		case 'd':	// English format input.
			opts.o_d = 1;
			opts.o_m = 0;
			opts.o_y = 0;
			break;
		case 'y':	// ISO format input.
			opts.o_y = 1;
			opts.o_m = 0;
			opts.o_d = 0;
			break;
		/* output options */
		case 'Y':	// to be ordered ymd
			opts.o_Y = 1;
			opts.o_D = 0;
			break;
		case 'D':	// to be ordered dmy
			opts.o_D = 1;
			opts.o_Y = 0;
			break;
		case 'r':	// real Thai numbers, not ASCII digits.
			opts.o_r = 1;
			break;
		case ':':
			fprintf(stderr, "Option %s requires an argument\n",
					argv[this_option_optind]);
			dohelp(1);
		break;
		case '?':
			fprintf(stderr, "Unknown option: %s\n",
					 argv[this_option_optind]);
			dohelp(1);
		break;
		} // switch()
	} // while()
	return opts;
} // process_options()

void dohelp(int forced)
{/* invoke manpage instead of trying to keep 2 sets of text synced. */
	char command[NAME_MAX];
	// so where is the manpage?
	if (exists_file("./thaidate.1")) {	// development time.
		sprintf(command, "man ./thaidate.1");
	} else {
		sprintf(command, "man thaidate.1");
	}
	xsystem(command, 1);
	exit(forced);
} // dohelp()


build.sh
#!/bin/bash
#
# build.sh - script to build a C program, options allow optimisation
#            or build for `gdb`.


# set defaults
debug=1;
verbose=0;

# write actual usage
usage () {
	echo;
	echo "./build.sh [option]"; echo;
	echo options;
	echo -h print help message and exit;
	echo '-d generate debug code (default)';
	echo '-o generate optimised code (O2)';
}

# options string
options=':ohd'
# the leading ':' in options string is required for the errors cases.

while getopts $options option
do
	case $option in
		o  ) debug=0;;
		d  ) debug=1;;
		h  ) usage; exit;;
		\? ) echo "Unknown option: -$OPTARG" >&2; exit 1;;
		:  ) echo "Missing option argument for -$OPTARG" >&2; exit 1;;
		*  ) echo "Unimplemented option: -$OPTARG" >&2; exit 1;;
	esac
done

if [ $debug == 1 ]
then
	gcc -Wall -Wextra -g -O0 -c -D_GNU_SOURCE=1 thaidate.c
	gcc -Wall -Wextra -g -O0 -c -D_GNU_SOURCE=1 files.c
	gcc -Wall -Wextra -g -O0 -c -D_GNU_SOURCE=1 gopt.c
	gcc -Wall -Wextra -g -O0 -c -D_GNU_SOURCE=1 str.c
else
	gcc -Wall -Wextra -O2 -c -D_GNU_SOURCE=1 thaidate.c
	gcc -Wall -Wextra -O2 -c -D_GNU_SOURCE=1 files.c
	gcc -Wall -Wextra -O2 -c -D_GNU_SOURCE=1 gopt.c
	gcc -Wall -Wextra -O2 -c -D_GNU_SOURCE=1 str.c
fi

# link stage
gcc thaidate.o files.o gopt.o str.o -o thaidate
rm *.o


thaidate.c
/*    thaidate.c
 *
 * Copyright 2017 Robert L (Bob) Parker rlp1938@gmail.com
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
*/

#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <fcntl.h>
#include <stdarg.h>
#include <getopt.h>
#include <ctype.h>
#include <limits.h>
#include <linux/limits.h>
#include <libgen.h>
#include <errno.h>
// typdefs/structs here.
typedef struct progdata {
	int dayin;
	int monin;
	int yerin;
	int yerout;
	int wantreal;
	int oporder;	// 0 = default, dd mmm yyyy else yyyy mmm dd
} progdata;

#include "str.h"
#include "files.h"
#include "gopt.h"

static progdata
*validate_input(options_t *opts, char **argv);
static char
*generateop(progdata *pd);
static char
*genthaidigits(int num);


int main(int argc, char **argv)
{
	options_t opts = process_options(argc, argv);
	progdata *pd = validate_input(&opts, argv);
	char *opstr = generateop(pd);
	fprintf(stdout, "%s\n", opstr);
	free(pd);	//nothing more to do, no pointers inside the struct.
	return 0;
}//main()

progdata
*validate_input(options_t *opts, char ** argv)
{/* All being well return valid progdata* */
	// First up make sure all args are there.
	if (!argv[optind] || !argv[optind+1] || !argv[optind+2]) {
		fputs("Must have day month and year as separate fields.\n",
		stderr);
		exit(EXIT_FAILURE);
	}
	int one, two, three, y, m, d;
	one = strtol(argv[optind], NULL, 10);
	optind++;
	two = strtol(argv[optind], NULL, 10);
	optind++;
	three = strtol(argv[optind], NULL, 10);
	if (opts->o_y) {	// ISO input
		y = one;
		m = two;
		d = three;
	} else if (opts->o_d) {	// dmy input;
		d = one;
		m = two;
		y = three;
	} else {	// mdy input
		m = one;
		d = two;
		y = three;
	}
	// test month first.
	if (m < 1 || m > 12) {
		fprintf(stderr, "Invalid month input: %d\n", m);
		exit(EXIT_FAILURE);
	}
	// validate year.
	int leap = (y % 4 == 0);
	leap = (leap && !(y % 100 == 0));
	leap = (leap || (y % 400 == 0));
	// validate day in month.
	int dim[13] = {-1, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
	if (leap) dim[2] = 29;
	if (!(d > 0 && d <= dim[m])) {
		fprintf(stderr, "Invalid days: %d for month: %d\n", d, m);
		exit(EXIT_FAILURE);
	}
	progdata *res = xmalloc(sizeof(progdata));
	res->dayin = d;
	res->monin = m;
	res->yerin = y;
	res->yerout = y + 543;
	res->wantreal = opts->o_r;
	res->oporder = opts->o_Y;	// 1 is yyyy mmm dd, 0 dd mmm yyyy
	return res;
} // validate_input()

static char
*generateop(progdata *pd)
{/* build the output string */
	char *monamesth[] = {"none", "มกราคม", "กุมภาพันธ์", "มีนาคม", "เมษายน",
						"พฤษภาคม", "มิถุนายน", "กรกฎาคม", "สิงหาคม",
						"กันยายน", "ตุลาคม", "พฤศจิกายน", "ธันวาคม"};
	// has max length 27 (november). 3 chars per visible char.
	char thmoname[32];
	strcpy(thmoname, monamesth[pd->monin]);
	char daysth[8];		// 2 days * 3 + '\0' is 7 needed.
	char yearth[16];	// 4 digits * 3 + '\0' is 13 needed.
	if (pd->wantreal) {	// buffers have to be bigger
		char *cp = genthaidigits(pd->dayin);
		strcpy(daysth, cp);
		cp = genthaidigits(pd->yerout);
		strcpy(yearth, cp);
	} else {
		sprintf(daysth, "%d", pd->dayin);
		sprintf(yearth, "%d", pd->yerout);
	}
	static char outbuf[PATH_MAX];
	// outut order ymd or dmy
	if (pd->oporder) {	// ymd
		sprintf(outbuf, "%s %s %s", yearth, thmoname, daysth);
	} else {	// dmy
		sprintf(outbuf, "%s %s %s", daysth, thmoname, yearth);
	}
	return outbuf;
}

char
*genthaidigits(int num)
{/* given an abitrary integer, return a Thai digit string */
	char *digitnamesth[] = {"๐", "๑", "๒", "๓", "๔", "๕", "๖", "๗",
						   /*0    1    2    3    4    5    6    7*/
							"๘", "๙"};
						   /*8    9 */
	static char buffer[NAME_MAX];
	buffer[0] = 0;
	sprintf(buffer, "%d", num);
	int l = strlen(buffer);
	// sanity check
	if (3 * l + 1 > NAME_MAX) {
		fprintf(stderr, "Your number %d, is too large.\n", num);
		exit(EXIT_FAILURE);
	}
	int i;
	int exp = 1;
	for (i = 0; i < l; i++) {
		exp *= 10;
	}
	exp /= 10;	// a digit too long
	int divby = exp;
	buffer[0] = 0;
	for (i = 0; i < l; i++) {
		int dd = num / divby;
		strcat(buffer, digitnamesth[dd]);
		num %= divby;
		divby /= 10;
	}
	return buffer;
}


gopt.h
/*      gopt.h
 *
 *  Copyright 2017 Robert L (Bob) Parker rlp1938@gmail.com
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *  MA 02110-1301, USA.
*/

#ifndef GOPT_H
#define GOPT_H
char *optstring;

typedef struct options_t {	// to be initialised with required vars.
	int	 o_m;	// -m, --mdy
	int	 o_d;	// -d, --dmy
	int	 o_y;	// -y, --ymd
	int	 o_Y;	// -Y, --YMD
	int	 o_D;	// -D, --DMY
	int	 o_r;	// -r, --real-thai

} options_t;

void dohelp(int forced);
options_t process_options(int argc, char **argv);

#endif
